Panic in pipeline.MarshalRESP

We're running commit:
```
github.com/mediocregopher/radix (git) - a2b04d3 Merge pull request #139 from nussjustin/fix-implicit-pipelining-net-error-handling
```
When we started our application it immediately panic'd with:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xb213ce]
goroutine 275 [running]:
github.com/mediocregopher/radix.(*pipelinerCmd).MarshalRESP(0xc000473c40, 0x12d3280, 0xc0001c95b0, 0xc000016cf8, 0x4138a5)
<autogenerated>:1 +0x2e
github.com/mediocregopher/radix.pipeline.MarshalRESP(0xc0001c9790, 0x1, 0x1, 0x12d3280, 0xc0001c95b0, 0x18, 0xc0005869a0)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/action.go:466 +0x72
github.com/mediocregopher/radix.(*connWrap).Encode(0xc000586640, 0x12d66c0, 0xc0005869a0, 0xc0005869a0, 0xc000016e20)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/radix.go:239 +0x4f
github.com/mediocregopher/radix.(*ioErrConn).Encode(0xc000586660, 0x12d66c0, 0xc0005869a0, 0xc0005869a0, 0x1b93428)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pool.go:40 +0x5c
github.com/mediocregopher/radix.pipelinerPipeline.Run(0xc0001c9790, 0x1, 0x1, 0x12f5d20, 0xc000586660, 0x23fe5b85, 0x291817668a2057)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pipeliner.go:225 +0xa8
github.com/mediocregopher/radix.(*ioErrConn).Do(...)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pool.go:61
github.com/mediocregopher/radix.(*Pool).Do(0xc000164d00, 0x12e1360, 0xc000586980, 0xc000586980, 0x1)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pool.go:607 +0xf0
github.com/mediocregopher/radix.(*pipeliner).flush.func1(0xc000442d80, 0xc0001c9790, 0x1, 0x1)
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pipeliner.go:182 +0x108
created by github.com/mediocregopher/radix.(*pipeliner).flush
/opt/TeamCityBuild/work/rachel-2/524c221af2f6e433/.deps/src/github.com/mediocregopher/radix/pipeliner.go:173 +0x8f
```

Since this happened during initialization I have to imagine its caused by something with:
```
df := func(network, addr string) (radix.Client, error) {
	...
	return radix.NewPool(network, addr, poolSize,
			radix.PoolConnFunc(func(network, addr string) (radix.Conn, error) {
				return radix.Dial(network, addr,
					radix.DialTimeout(timeout),
				)
			}),
			radix.PoolOnEmptyCreateAfter(100*time.Millisecond),
			radix.PoolOnFullBuffer(poolSize*10, time.Second),
		)
	}
}
...
conn, err := df("tcp", addr)
if err != nil {
  return nil, err
}
defer conn.Close()

var c radix.Client
err = conn.Do(radix.Cmd(nil, "CLUSTER", "SLOTS"))
if err == nil {
  c, err = radix.NewCluster([]string{ar.Addr},
    radix.ClusterPoolFunc(df),
  )
} else if isClusterNotSupportedError(err) {
  c, err = df("tcp", ar.Addr)
}
```

I have to imagine its something with `defer conn.Close()` and that somehow something is still trying to use that after its closed but I'm not actually sure that it ever got past `err = conn.Do(radix.Cmd(nil, "CLUSTER", "SLOTS"))` so that could also be the issue.

--- a/pipeliner.go
+++ b/pipeliner.go
@@ -181,10 +181,10 @@ func (p *pipeliner) flush(reqs []CmdAction) []CmdAction {
 
 		if err := p.c.Do(pipe); err != nil {
 			for _, req := range reqs {
-				select {
-				case req.(*pipelinerCmd).resCh <- err:
-				default:
+				if req == nil {
+					continue
 				}
+				req.(*pipelinerCmd).resCh <- err
 			}
 		}
 	}()
@@ -226,7 +226,7 @@ func (p pipelinerPipeline) Run(c Conn) error {
 		return err
 	}
 	errConn := ioErrConn{Conn: c}
-	for _, req := range p.pipeline {
+	for i, req := range p.pipeline {
 		err := errConn.Decode(req)
 		// the order here is important: if we tried to send the error to
 		// before returning we could end up sending the error twice, which
@@ -237,6 +237,12 @@ func (p pipelinerPipeline) Run(c Conn) error {
 			return errConn.lastIOErr
 		}
 		req.(*pipelinerCmd).resCh <- err
+		// since the *pipelinerCmd can be pooled and reused right after
+		// we send the response, even before this goroutine returns from
+		// the send operation, we must not access it again. to make sure
+		// that we don't regress anywhere we set it to nil, so that the
+		// runtime will trigger a panic if we try to access cmd again.
+		p.pipeline[i] = nil
 	}
 	return nil
 }

--- a/pipeliner.go
+++ b/pipeliner.go
@@ -181,7 +181,10 @@ func (p *pipeliner) flush(reqs []CmdAction) []CmdAction {
 
 		if err := p.c.Do(pipe); err != nil {
 			for _, req := range reqs {
-				req.(*pipelinerCmd).resCh <- err
+				select {
+				case req.(*pipelinerCmd).resCh <- err:
+				default:
+				}
 			}
 		}
 	}()

--- a/pipeliner.go
+++ b/pipeliner.go
@@ -227,10 +227,11 @@ func (p pipelinerPipeline) Run(c Conn) error {
 	}
 	errConn := ioErrConn{Conn: c}
 	for _, req := range p.pipeline {
-		req.(*pipelinerCmd).resCh <- errConn.Decode(req)
+		err := errConn.Decode(req)
 		if errConn.lastIOErr != nil {
 			return errConn.lastIOErr
 		}
+		req.(*pipelinerCmd).resCh <- err
 	}
 	return nil
 }

--- a/pipeliner.go
+++ b/pipeliner.go
@@ -225,6 +225,11 @@ func (p pipelinerPipeline) Run(c Conn) error {
 	errConn := ioErrConn{Conn: c}
 	for _, req := range p.pipeline {
 		err := errConn.Decode(req)
+		// the order here is important: if we tried to send the error to
+		// before returning we could end up sending the error twice, which
+		// could lead to data races when the *pipelinerCmd gets reused.
+		// to avoid this we must return without sending a value, in case
+		// we got a non-recoverable error.
 		if errConn.lastIOErr != nil {
 			return errConn.lastIOErr
 		}

--- a/pipeliner.go
+++ b/pipeliner.go
@@ -181,10 +181,7 @@ func (p *pipeliner) flush(reqs []CmdAction) []CmdAction {
 
 		if err := p.c.Do(pipe); err != nil {
 			for _, req := range reqs {
-				select {
-				case req.(*pipelinerCmd).resCh <- err:
-				default:
-				}
+				req.(*pipelinerCmd).resCh <- err
 			}
 		}
 	}()

